[Inductor] Fix CPU vectorized implementation of mask calculation that breaks torch.where

jgong5 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #93922

Fix https://github.com/pytorch/pytorch/issues/93374

The cause of the issue is that the original vectorized float mask calculation doesn't consider the broadcast case. This PR adds the support.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Commit:
[Inductor] Fix CPU vectorized implementation of torch.where

[ghstack-poisoned]

Commit:
Update on "[Inductor] Fix CPU vectorized implementation of torch.where"


Fix https://github.com/pytorch/pytorch/issues/93374

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

Commit:
Update on "[Inductor] Fix CPU vectorized implementation of torch.where"


Fix https://github.com/pytorch/pytorch/issues/93374

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp_prefix.h:
```diff
@@ -69,3 +69,12 @@ void flag_to_float(const T* src, float* dst, int64_t n) {
 }
+
+template <typename T>
+void flag_to_float(T src, float* dst, int64_t n) {
```
This overload is needed to handle the broadcasting case where vectorized float mask construction with a scalar.

desertfire reviewed (approved):
Thanks for the quick fix.

jgong5 commented:
@pytorchbot merge

Commit:
Update on "[Inductor] Fix CPU vectorized implementation of mask calculation that breaks torch.where"


Fix https://github.com/pytorch/pytorch/issues/93374

The cause of the issue is that the original vectorized float mask calculation doesn't consider the broadcast case. This PR adds the support.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented:
@pytorchbot merge
